raiden/utils/upgrades.py

```diff
@@ -0,0 +1,127 @@
+import os
+import shutil
+import sqlite3
+from pathlib import Path
+
+import filelock
+import sqlitebck
+import structlog
+
+from raiden.exceptions import RaidenDBUpgradeError
+from raiden.storage.serialize import JSONSerializer
+from raiden.storage.sqlite import RAIDEN_DB_VERSION, SQLiteStorage
+from raiden.storage.versions import older_db_file
+from raiden.utils.migrations.v16_to_v17 import upgrade_initiator_manager
+
+UPGRADES_LIST = [
+    upgrade_initiator_manager,
+]
+
+
+log = structlog.get_logger(__name__)
+
+
+def get_file_lock(db_filename: Path):
+    lock_file_name = f'{db_filename}.lock'
+    return filelock.FileLock(lock_file_name)
+
+
+def get_db_version(db_filename: Path):
+    db_lock = get_file_lock(db_filename)
+    with db_lock:
+        # Perform a query directly through SQL rather than using
+        # storage.get_version()
+        # as get_version will return the latest version if it doesn't
+        # find a record in the database.
+        conn = sqlite3.connect(
+            str(db_filename),
+            detect_types=sqlite3.PARSE_DECLTYPES,
+        )
+        cursor = conn.cursor()
+        try:
+            cursor.execute('SELECT value FROM settings WHERE name=?;', ('version',))
+            query = cursor.fetchall()
+            if len(query) == 0:
+                return 0
+            return int(query[0][0])
+        except sqlite3.OperationalError:
```

Comment:
This should not be handled, the exception should propagate and kill the client:

> Exception raised for errors that are related to the database’s operation and not necessarily under the control of the programmer

https://docs.python.org/3/library/sqlite3.html#sqlite3.OperationalError

Reply:
So on startup, we try to get the current version's DB from the `settings` table. If you just ran a new Raiden version, say version 17 while you were on 16 before that, the DB file will be empty... no tables will exist there. Therefore, we try to get the version (if the database is valid)... if it's not (no tables exist on a brand new file) then an `OperationalError` will be thrown, by that time we know that we have a new file that hasn't been populated with tables yet so we return 0.